### ISSUE TITLE
Could not create SSL/TLS secure

### ISSUE THREAD

**Issue by amitabgwebs** [adplabs/adp-connection-NET#13]
I am getting this error when project updated on azure server for testing. Its working on localserver.

Getting error on this code
 connection.connect();

Error message:
An error occurred while sending the request. ---> System.Net.WebException: The request was aborted: Could not create SSL/TLS secure channel. at System.Net.HttpWebRequest.EndGetRequestStream(

**Comment by gregarican** [comment]
My guess without looking into specifics is perhaps you’re using a deprecated TLS version? Such as 1.0 versus 1.2 versus 2.0.

Sent from my iPhone

On Jul 19, 2018, at 9:40 AM, amitabgwebs <notifications@github.com<mailto:notifications@github.com>> wrote:


I am getting this error when project updated on azure server for testing. Its working on localserver.

Getting error on this code
connection.connect();

Error message:
An error occurred while sending the request. ---> System.Net.WebException: The request was aborted: Could not create SSL/TLS secure channel. at System.Net.HttpWebRequest.EndGetRequestStream(

**Comment by amitabgwebs** [comment]
I checked and could't find TLS2.0 I am using TLS 1.2 in my code.

Not sure if I am correct,but I am using below code before connection.

```
                    ServicePointManager.Expect100Continue = true;
                    ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;
                   connection.connect();
```

Any help is appreciated, thanks.

**Comment by migdiada** [comment]
Haven't changed anything. Am getting this error suddenly on Tuesday
afternoon using the same code that's been working for a month.

We have 3 instances that are not working, while, if I use Postman to
manually call the API, it does work. Using the same certs in all places.
Even reduced the code to just a basic workers api call and get a 401 using
the python code (from adp) that we've been using all along.

Am trying to determine what it could be when the python connect lib stops
working but postman works (and even taking the code that python is using
for requests and that is the exact code I'm using in the python script. The
authentication using certs and api keys, etc.. IS working because I am
getting a successful connection in python, and getting a valid Bearer
token, after connecting.

The 401 happens when I call the workers API.

Thanks for any suggestions. If we could hop on a call.. would love anyone
to look over my shoulder.. that sometimes helps.

Thanks!

Russ

**Comment by gregarican** [comment]
All I recall is that this summer I had to explicitly set the correct TLS version for a good number of larger third party .NET API’s. Since they deprecated TLS to require 2.0. Postman worked while troubleshooting since it invoked the latest version by default. Look at Fiddler traces and you can see which TLS version your code is referencing.
Worked for me!

Sent from my iPhone

On Jul 19, 2018, at 10:41 AM, rhf <notifications@github.com<mailto:notifications@github.com>> wrote:

Haven't changed anything. Am getting this error suddenly on Tuesday
afternoon using the same code that's been working for a month.

We have 3 instances that are not working, while, if I use Postman to
manually call the API, it does work. Using the same certs in all places.
Even reduced the code to just a basic workers api call and get a 401 using
the python code (from adp) that we've been using all along.

Am trying to determine what it could be when the python connect lib stops
working but postman works (and even taking the code that python is using
for requests and that is the exact code I'm using in the python script. The
authentication using certs and api keys, etc.. IS working because I am
getting a successful connection in python, and getting a valid Bearer
token, after connecting.

The 401 happens when I call the workers API.

Thanks for any suggestions. If we could hop on a call.. would love anyone
to look over my shoulder.. that sometimes helps.

Thanks!

Russ

**Comment by migdiada** [comment]
Thank you so much!! I will try that! Very greatful for your advice. I will
let you know!

**Comment by migdiada** [comment]
not sure if I mentioned.. I am using python

**Comment by sloppypointless** [comment]
Have you checked the expiration date of your counter signed ADP SSL Certificate?

FYI TLS 2.0 doesn't exist.  

I suspect he meant to make sure you aren't using one of the deprecated versions which I suspect are SSL 1.0, SSL 2.0, SSL 3.0, TLS 1.0 and TLS 1.1.

TLS 1.2 is the most recent widely accepted version.

TLS 1.3 is still a proposed specification at this time I believe.

**Comment by gregarican** [comment]
Correct! Sorry about that. Was mixing up SSL versions with TLS versions. Various third party API web services I connect to for my production apps required TLS 1.2, as TLS 1.0 and 1.1 were retired due to widespread security initiatives most companies have been taking this summer. By default my Visual Studio and .NET references weren’t invoking TLS 1.2 by default, so I had to explicitly specify it. Not sure if Python might be doing something similar under the hood…

**Comment by ADPMarketplace** [comment]
Hi,
We tested the code which is there in GitHub and it's working. Could you please check if you're passing the correct certificates while calling ADP API. Please download the python-code once again and test it and let us know if you still get the same issue.

Thanks,
Mahesh

**Comment by migdiada** [comment]
I resolved the issue

**Comment by gregarican** [comment]
So that we all have better closure as to the issue, what did you do that resolved it?

Greg Kujawa
IT Manager

DIAMOND CELLAR Holdings, LLC
6280 Sawmill Road
Dublin, OH 43017
614-923-9515 phone
614-336-4555 fax
www.diamondcellar.com<http://www.diamondcellar.com>
Office Hours: M Tu W Th F

[cid:image002.png@01CF8981.56A564A0]<https://www.facebook.com/DiamondCellar>[cid:image003.png@01CF8981.56A564A0]<https://twitter.com/DiamondCellarCo>[cid:image004.png@01CF8981.56A564A0]<http://www.pinterest.com/diamondcellar/>[cid:image005.png@01CF8981.56A564A0]<http://www.yelp.com/biz/diamond-cellar-dublin>[cid:image006.png@01CF8981.56A564A0]<http://instagram.com/diamondcellar>[cid:image007.png@01CF8981.56A564A0]<https://plus.google.com/+DiamondCellarDublin/about>


CONFIDENTIALITY NOTICE
UNAUTHORIZED INTERCEPTION IS PROHIBITED BY FEDERAL LAW
[Electronic Communications Privacy Act of 1986, 18 U.S.C. 2701(a) and 2702(a)] This message is protected by Constitutional Rights and applicable legal privileges.  It is strictly confidential.

**Comment by ADPMarketplace** [comment]
Hi Kujawa,
No change we did at our end. Not sure, may be at your end, you're not passing the certificates properly earlier.

As the issue is resolved now. Shall I close this ticket?

Thanks,
Mahesh

**Comment by migdiada** [comment]
My app is running fine now. There was some communication that went out that
I did not receive that the API would start requiring the cert with every
API call, not just the token request.

**Comment by bplus0** [comment]
> Hi,
> We tested the code which is there in GitHub and it's working. Could you please check if you're passing the correct certificates while calling ADP API. Please download the python-code once again and test it and let us know if you still get the same issue.
> 
> Thanks,
> Mahesh

Where in the github code can you see an example of this? I'm getting the same exact error in my .net application, and i cannot figure out why.

It works perfect locally, but when it's published to the server. tells me error with TLS. Can you please help?

**Comment by gregarican** [comment]
At this point this is a pretty old thread. Most providers retiring TLS 1.0 and 11. is old news. The two things that need to be done involve:

1. Ensuring that your code has explicitly specified **System.Net.ServicePointManager.SecurityProtocol = System.Net.SecurityProtocolType.Tls12;** in the main application startup.
2. Ensuring that the host you are running the code on is configured for TLS 1.2 at the lowest.

**Comment by bplus0** [comment]
Thanks. I have been on that thread.. I can't find the exact place you need to tell the app to use your certificate.

It was working on server for us, then one day it stopped.
